# PYSPARK-2_Mandatory_Project

# Introduction :

The Bellow data set is all about the details of Airlines the detailed explanation is as follows.

-> The columns like Year,Month,Day,Week says that on which date and the days that these all data Belongs to.

-> The columns like CarrierDelay,WeatherDelay,NAS_Delay,SecurityDelay,LateAircraftDelay says that on which reason that the flights were Delayed.

-> DepTime : Total Depature time.

-> CRSDepTime : If it is operated 15 minutes before the sheduled time.

-> ArrTime : Total Arrival time.

-> CRSArrTime : If it is operated 15 minutes later the sheduled time.

->  UniqueCarrier : Flights that each airline made.

-> FlightNum : Number of the particular flight.

-> ActualElapsedTime : The total time tha SLA has taken untill this time,minus any pause duration.

-> CRSElapsedTime : Difference in minutes between sheduled and actual arrival time.

-> ArrDelay : Arrival Delay of a flight.

-> DepDelay : Depature Delay of a flight.

-> Origin : For which origin that flight belongs to.

-> Distance : Total diatance.

-> Cancelled : Total Number of flights were cancled.

-> CancellationCode : what is the code of that cancelled flights.

-> Diverted : How many number of flights were Diverted.

In [1]:
pip install pyspark # installing pyspark

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 281.3 MB 46 kB/s 
     |████████████████████████████████| 199 kB 52.7 MB/s 
  Created wheel for pyspark: filename=pyspark-3.3.0-py2.py3-none-any.whl size=281764026 sha256=682178f3f27c01c1b0f6dfac60289bca243f51bcc0e3b204007f175b8af110d4
  Stored in directory: /root/.cache/pip/wheels/7a/8e/1b/f73a52650d2e5f337708d9f6a1750d451a7349a867f928b885
Successfully built pyspark


In [2]:
#Initializing PySpark and importing the other required libraries
from pyspark import SparkContext, SparkConf
from pyspark.sql import SparkSession
from pyspark.sql import SQLContext
import pyspark.sql.functions as f

#Spark Config
conf = SparkConf().setAppName("sample_app")
sc = SparkContext(conf=conf)
spark = SparkSession.builder.appName('Test').getOrCreate()
sqlContext = SQLContext(sc)

/usr/local/lib/python3.7/dist-packages/pyspark/sql/context.py:114: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
  FutureWarning,


In [25]:
# reading the data set of csv file to a Data Frame
airline_data = spark.read.format("csv").option("header","true").option("inferschema","true").load("/content/Airline_data.csv")
airline_data.show() # calling the Action

+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+
|Year|Month|DayofMonth|DayOfWeek|DepTime|CRSDepTime|ArrTime|CRSArrTime|UniqueCarrier|FlightNum|TailNum|ActualElapsedTime|CRSElapsedTime|AirTime|ArrDelay|DepDelay|Origin|Dest|Distance|TaxiIn|TaxiOut|Cancelled|CancellationCode|Diverted|CarrierDelay|WeatherDelay|NASDelay|SecurityDelay|LateAircraftDelay|
+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+
|1989|    1|        23|        1|   1419|      1230|   1742|      1552|           UA|      183

# Questions :

## Q1) Show a sample of 5 records from dataset.

In [26]:
sample_data = airline_data.limit(5) # getting the sample of airline_data by using "limit" function
sample_data.show() # calling the Action

+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+
|Year|Month|DayofMonth|DayOfWeek|DepTime|CRSDepTime|ArrTime|CRSArrTime|UniqueCarrier|FlightNum|TailNum|ActualElapsedTime|CRSElapsedTime|AirTime|ArrDelay|DepDelay|Origin|Dest|Distance|TaxiIn|TaxiOut|Cancelled|CancellationCode|Diverted|CarrierDelay|WeatherDelay|NASDelay|SecurityDelay|LateAircraftDelay|
+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+
|1989|    1|        23|        1|   1419|      1230|   1742|      1552|           UA|      183

## Explanation :
  ###  As asked in the Question Displayed the sample of five records By using limit function. The limit function generally limits the No of rows to display.

## Q2) Read the data with data types.

In [5]:
# Reading the data types by using the function called printSchema
data_types = airline_data.printSchema() 
data_types # calling the Action

root
 |-- Year: integer (nullable = true)
 |-- Month: integer (nullable = true)
 |-- DayofMonth: integer (nullable = true)
 |-- DayOfWeek: integer (nullable = true)
 |-- DepTime: integer (nullable = true)
 |-- CRSDepTime: integer (nullable = true)
 |-- ArrTime: integer (nullable = true)
 |-- CRSArrTime: integer (nullable = true)
 |-- UniqueCarrier: string (nullable = true)
 |-- FlightNum: integer (nullable = true)
 |-- TailNum: string (nullable = true)
 |-- ActualElapsedTime: integer (nullable = true)
 |-- CRSElapsedTime: integer (nullable = true)
 |-- AirTime: string (nullable = true)
 |-- ArrDelay: integer (nullable = true)
 |-- DepDelay: integer (nullable = true)
 |-- Origin: string (nullable = true)
 |-- Dest: string (nullable = true)
 |-- Distance: integer (nullable = true)
 |-- TaxiIn: string (nullable = true)
 |-- TaxiOut: string (nullable = true)
 |-- Cancelled: integer (nullable = true)
 |-- CancellationCode: string (nullable = true)
 |-- Diverted: integer (nullable = true)
 |

## Explanation :
   ### There Two data types are present in the above data set those are 1) Integer, 2) String.

## Integer columns are :
### Year,Month,DayofMonth,DayofWeek,DepTime,CESDepTime,ArrTime,CESArrTime,FlightNumber,ActualElapsedTime,CRSelapsedTime,ArrDelay,DepDelay,Distance,Cancelled,Diverted.

## String columns are :
### UniqueCarrier,Tailnum,AirTime,Origin,Dest,TaxiIn,taxiOut,CancellationCode,WeatherDealy,NASDelay,SecurityDelay,LateAircraftDelay.



# NOTE :
### In the above Data set some integer columns are Reading as strings because of its Null values.

## Q3) Make a new column MonthStr, Which has months in form of 01, 02, 03, ..., 12.

In [27]:
# program for tne required condition 
def new_col(x):
    if x == 1:
      return '01'
    elif x == 2:
      return '02'
    elif x == 3:
      return'03'
    elif x == 4:
      return'04'
    elif x == 5:
      return'05'
    elif x == 6:
      return '06'
    elif x == 7:
      return '07'
    elif x == 8:
      return '08'
    elif x == 9:
      return '09'
    else:
      return x

In [28]:
# importing udf & col to add a new colum with a condition to a PYSPARK DataFrame
from pyspark.sql.functions import udf,col

In [29]:
a = udf(new_col) # creting UDF to apply  the above programed condition to the pyspark dataframe
new_col = airline_data.withColumn("MonthStr",a(col("Month"))) # creating a new column with required condition
new_col.show() # calling the Action

+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+--------+
|Year|Month|DayofMonth|DayOfWeek|DepTime|CRSDepTime|ArrTime|CRSArrTime|UniqueCarrier|FlightNum|TailNum|ActualElapsedTime|CRSElapsedTime|AirTime|ArrDelay|DepDelay|Origin|Dest|Distance|TaxiIn|TaxiOut|Cancelled|CancellationCode|Diverted|CarrierDelay|WeatherDelay|NASDelay|SecurityDelay|LateAircraftDelay|MonthStr|
+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+--------+
|1989|    1|        23|        1|   1419|      1230|   1742|      1

In [30]:
# Checking whether the required condition was applied to new column or not 
check = new_col.select("Month","MonthStr")
check.show(5) # calling the Action

+-----+--------+
|Month|MonthStr|
+-----+--------+
|    1|      01|
|    1|      01|
|    1|      01|
|    1|      01|
|    1|      01|
+-----+--------+
only showing top 5 rows



## Explanation :
  ### Here we can clearly see that in the month column the data was chaged as per the requirement.Here all the values in the newly created column was "01" only because in the original Month column all the values are "1" only.

## Q4) Find the number of flights each airline made.

In [31]:
airline_data.createOrReplaceTempView("airline") # creating a view to write an SQL
Numner_of_flights = spark.sql("select count(UniqueCarrier) from airline ") # counting No of flights each airline made
Numner_of_flights.show() # calling the Action

+--------------------+
|count(UniqueCarrier)|
+--------------------+
|                 426|
+--------------------+



## Explanation :
 ### The above result clearly showing that the number of flights that each airline made is 426

## Q5) Find the mean Arrival Delay per origination airport.

In [32]:
# Grouping and averaging the Required columns as per the Question
mean_arr_delay_per_org = airline_data.groupBy("Origin").avg("ArrDelay")
mean_arr_delay_per_org.show() # Calling the Action

+------+-------------------+
|Origin|      avg(ArrDelay)|
+------+-------------------+
|   LIH|0.16666666666666666|
|   HNL|  14.21774193548387|
|   EWR|               9.25|
|   DEN| 20.166666666666668|
|   IAD| 12.966666666666667|
|   SFO| 11.215384615384615|
|   PHL|  6.827586206896552|
|   OGG|  16.24137931034483|
+------+-------------------+



## Explanation :
### The above result clearly showing the details of average number of Arrival Delays of flights in each Origin.

### By above result we can say that Origin "DEN" has most  Number of Arrival Delays with average arrival delay of (20.166666666666668) where as Origin "LIH" has less number of Arrival Delays with average arrival delay of (0.16666666666666666).

## Q6)  What is the average departure delay from each airport.

In [33]:
# Grouping and averaging the Required columns as per the Question
avg_dep_delay_per_airport = airline_data.groupBy("Origin").avg("DepDelay")
avg_dep_delay_per_airport.show() # Calling the Action

+------+-------------------+
|Origin|      avg(DepDelay)|
+------+-------------------+
|   LIH|-3.7666666666666666|
|   HNL|  3.217741935483871|
|   EWR|  4.958333333333333|
|   DEN|               27.6|
|   IAD|                8.9|
|   SFO| 19.646153846153847|
|   PHL| 16.137931034482758|
|   OGG|                6.0|
+------+-------------------+



## Explanation :
### The above result clearly showing the details of average number of Depature Delays of flights in each Origin.

### By above result we can say that Origin "DEN" has most  Number of Depature Delays with average arrival delay of (27.6) where as Origin "LIH" has less number of Depature Delays with average Depature delay of (-3.7666666666666666).